### ISSUE TITLE
Crash on shader compilation and project loading (100% reproducible)

### ISSUE THREAD

**Issue by viktor-ferenczi** [godotengine/godot#64948]
### Godot version

4.0.alpha14.official.106b68050

### System information

Windows 10, AMD Ryzen 3800X, 64GB RAM, nVidia RTX 3080Ti, Vulkan

### Issue description

Godot freezes right after extending a spatial shader code with enough varying flat global variables.

### Steps to reproduce

The problem is 100% reproducible by adding the following Shader to a simple cube in a brand new Vulkan project:

```
shader_type spatial;

uniform sampler2DArray sampler1;
uniform sampler2DArray sampler2;
uniform sampler2DArray sampler3;
uniform sampler2DArray sampler4;
uniform sampler2DArray sampler5;

const int A = 32;
const int B = A - 1;

varying flat vec3 VECTOR1;
varying flat mat4 MATRIX1;
varying flat mat4 MATRIX2;
varying flat mat3 MATRIX3;
varying flat mat3 MATRIX4;
varying flat mat3 MATRIX5;
varying flat mat3 MATRIX6;
varying flat mat3 MATRIX7;
varying flat mat3 MATRIX8;

void vertex() {
	VECTOR1 = vec3(0.0);
	MATRIX1 = mat4(1.0);
	MATRIX1 = mat4(1.0);
	MATRIX1 = mat4(1.0);
	MATRIX1 = mat4(1.0);
	MATRIX1 = mat4(1.0);
}

void fragment() {
}
```

Console log output while running with --debug (it crashes without the --debug option as well):

```
Godot Engine v4.0.alpha14.official.106b68050 - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce RTX 3080 Ti

Editing project: C:/Dev/Godot4/VulkanCrashOnShader

Godot Engine v4.0.alpha14.official.106b68050 - https://godotengine.org
ERROR: Condition "!windows.has(p_window)" is true.
   at: window_resize (drivers/vulkan/vulkan_context.cpp:1479)
Vulkan API 1.2.0 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce RTX 3080 Ti

WARNING: FBX file import is enabled, but no FBX2glTF path is configured. FBX files will not be imported.
     at: _editor_init (modules/gltf/register_types.cpp:92)
ERROR: Condition "err" is true. Returning: ERR_CANT_CREATE
   at: swap_buffers (drivers/vulkan/vulkan_context.cpp:2140)
ERROR: Condition "err" is true. Returning: ERR_CANT_CREATE
   at: swap_buffers (drivers/vulkan/vulkan_context.cpp:2140)
ERROR: Vulkan: Did not create swapchain successfully.
   at: (drivers/vulkan/vulkan_context.cpp:2063)
ERROR: Condition "err" is true. Returning: ERR_CANT_CREATE
   at: swap_buffers (drivers/vulkan/vulkan_context.cpp:2140)
```

The above log was on project loading. If I get this error while editing the shader, then the last error message repeats very rapidly and the whole Godot freezes until manually killed.

### Minimal reproduction project

Attached ZIP of the complete project to reproduce. It should freeze right on loading the project. Sometimes it gets into the UI from the splash screen, but the cube does not show up.

**Comment by clayjohn** [comment]
My guess is this shader is going over the device's varying limit ([maxVertexOutputComponents](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxVertexOutputComponents&platform=all)). It looks like its 128 for most devices, but the minimum required by the spec is 64. 

Most likely your device supports 128, which means it can occupy 32 locations. A vec3 takes 1, a Mat4 takes 4, and a Mat3 takes 3. Godot uses at least 2, but may use more depending on if the mesh has normals, UVs, etc.

This shader uses 27 locations, my guess is Godot is using up at least five more and you are running out of locations. 

If i am right, then the solution is to do more validation (i.e. check the number of locations used while compiling the shader to ensure the limit is not exceeded)

**Comment by viktor-ferenczi** [comment]
Thank you for the in depth information. I wasn't aware that it is limited this much. It lets me to work around the problem until more validation is added to shader compilation.

Do we have such a limit on the uniform variables as well?

**Comment by clayjohn** [comment]
> Do we have such a limit on the uniform variables as well?

Yes, Godot uses a single uniform buffer to store all user created uniforms. Most GPUs have a maximum uniform buffer size of 65536 bytes (https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxUniformBufferRange&platform=all). I believe we validate this as you write your shader, so you should get an in-editor warning if you exceed the limit.

**Comment by Anutrix** [comment]
![image](https://user-images.githubusercontent.com/11459028/188919882-0d99d096-9729-4319-bf79-8a046d42314d.png)
The MRP doesn't crash on load or show errors(except `ERROR: Condition "!windows.has(p_window)" is true.`). The cube shows up in preview.
Running the project also doesn't crash or show error. However, the cube doesn't show up in the game.
Editing the shader also did not cause crash.
Not sure when this crash might have been fixed or it might be a GPU issue.

```
Godot Engine v4.0.alpha.custom_build.3fe3c4399 - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: AMD - Radeon RX 570 Series
On Windows 11 running on AMD Ryzen 9 5900X
```
